src/full/Agda/TypeChecking/DisplayForm.hs

```diff
@@ -107,71 +109,84 @@ displayForm q es = do
 --   i.e., @es / ps = Just us@.
 matchDisplayForm :: MonadDisplayForm m
                  => DisplayForm -> Elims -> MaybeT m (DisplayForm, DisplayTerm)
-matchDisplayForm d@(Display _ ps v) es
+matchDisplayForm d@(Display n ps v) es
   | length ps > length es = mzero
   | otherwise             = do
       let (es0, es1) = splitAt (length ps) es
-      us <- reverse <$> do match ps $ raise 1 es0
+      mm <- match (Window 0 n) ps es0
+      us <- sequence [ do Just u <- return $ Map.lookup i mm; return u | i <- [0..n - 1] ]
       return (d, substWithOrigin (parallelS $ map woThing us) us v `applyE` es1)
 
+type MatchResult = Map Int (WithOrigin Term)
+
+unionMatch :: Monad m => MatchResult -> MatchResult -> MaybeT m MatchResult
+unionMatch m1 m2
+  | Set.null $ Set.intersection (Map.keysSet m1) (Map.keysSet m2) = return $ Map.union m1 m2
```

Comment:
Use `Set.disjoint`: https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Set.html#v:disjoint
(ok since we are on `containers >= 0.5.11.0`).